src/trace_queue.js

```diff
@@ -0,0 +1,188 @@
+/* eslint-disable prefer-destructuring */
+/* eslint-disable no-console */
+/* eslint-disable valid-jsdoc */
+/* eslint-disable require-jsdoc */
+const EventEmitter = require('events').EventEmitter;
+const axios = require('axios');
+const https = require('https');
+const http = require('http');
+const utils = require('../src/utils.js');
+const config = require('./config.js');
+const consts = require('./consts.js');
+
+
+class TraceQueue extends EventEmitter {
+    constructor(batchSender) {
+        super();
+        this.batchSize = config.getConfig().batchSize;
+        this.batchBytesLimit = consts.batchBytesLimit;
+        this.traces = [];
+        this.batchQueueTime = config.getConfig().batchQueueTime;
+        this.batchSender = batchSender;
+        this.currentByteSize = 0;
+        this.initQueue();
+    }
+
+    /**
+   * Init queue event listners
+   */
+    initQueue() {
+        this.removeAllListeners();
+        this.currentByteSize = 0;
+        this.flush();
+        //  check if queue reached batch size
+        this.on('traceQueued', () => {
```

Comment:
I think that using event emitter makes it so you can send a couple of batches in parallel - not sure this is what you want

Reply:
Yep, it doesn't seem like a good solution for this case.
We need that on reaching batch size / time the buffered traces will get sent asynchronously. Any other traces that are buffered should not be sent until sending the previous batch ends.

Reply:
That's relate to the releaseBatch comment below, let's talk about it

Reply:
@ronyis so the releaseBatch called Synced, while the sendBatch is Asynced